tree/ntuple/v7/test/ntuple_join_processor.cxx

```diff
@@ -0,0 +1,223 @@
+      EXPECT_THAT(err.what(), testing::HasSubstr("a maximum of four join fields is allowed"));
+   }
+
+   try {
+      RNTupleProcessor::CreateJoin(ntuples, {"i", "i"});
+      FAIL() << "trying to create a join processor with duplicate join fields should throw";
+   } catch (const RException &err) {
+      EXPECT_THAT(err.what(), testing::HasSubstr("join fields must be unique"));
+   }
+
+   auto proc = RNTupleProcessor::CreateJoin(ntuples, {"i"});
+
+   int nEntries = 0;
+   auto i = proc->GetEntry().GetPtr<int>("i");
+   auto x = proc->GetEntry().GetPtr<float>("x");
+   auto y = proc->GetEntry().GetPtr<std::vector<float>>("ntuple2.y");
+   std::vector<float> yExpected;
+   for (auto &entry : *proc) {
+      EXPECT_FLOAT_EQ(nEntries, *entry.GetPtr<int>("i"));
+
+      EXPECT_FLOAT_EQ(*i * 0.5f, *x);
+
+      if (*i % 2 == 1) {
+         yExpected = {static_cast<float>(*i * 0.2), 3.14, static_cast<float>(*i * 1.3)};
+         EXPECT_EQ(yExpected, *entry.GetPtr<std::vector<float>>("ntuple2.y"));
+      } else {
+         yExpected = {};
+         EXPECT_EQ(yExpected, *y);
+      }
+
+      ++nEntries;
+   }
+}
+
+TEST_F(RNTupleJoinProcessorTest, WithModel)
+{
+   auto model1 = RNTupleModel::Create();
+   auto i = model1->MakeField<int>("i");
+   auto x = model1->MakeField<float>("x");
+
+   auto model2 = RNTupleModel::Create();
+   auto y = model2->MakeField<std::vector<float>>("y");
+
+   auto model3 = RNTupleModel::Create();
+   auto z = model3->MakeField<float>("z");
+
+   std::vector<std::unique_ptr<RNTupleModel>> models;
+   models.push_back(std::move(model1));
+   models.push_back(std::move(model2));
+   models.push_back(std::move(model3));
+
+   std::vector<RNTupleOpenSpec> ntuples = {
+      {fNTupleNames[0], fFileNames[0]}, {fNTupleNames[1], fFileNames[1]}, {fNTupleNames[2], fFileNames[2]}};
+
+   auto proc = RNTupleProcessor::CreateJoin(ntuples, {"i"}, std::move(models));
+
+   std::vector<float> yExpected;
+   int nEntries = 0;
+   for (auto &entry : *proc) {
+      EXPECT_EQ(nEntries, *i);
+      EXPECT_EQ(*entry.GetPtr<int>("i"), *i);
+
+      EXPECT_FLOAT_EQ(*i * 0.5f, *x);
+      EXPECT_FLOAT_EQ(*entry.GetPtr<float>("x"), *x);
+
+      if (*i % 2 == 1) {
+         yExpected = {static_cast<float>(*i * 0.2), 3.14, static_cast<float>(*i * 1.3)};
+         EXPECT_EQ(yExpected, *y);
+         EXPECT_EQ(*entry.GetPtr<std::vector<float>>("ntuple2.y"), *y);
+         EXPECT_FLOAT_EQ(static_cast<float>(*i * 2.f), *z);
+         EXPECT_FLOAT_EQ(*entry.GetPtr<float>("ntuple3.z"), *z);
+      } else {
+         yExpected = {};
+         EXPECT_EQ(yExpected, *y);
+         EXPECT_EQ(0., *z);
+      }
+
+      try {
+         entry.GetPtr<float>("ntuple2.z");
+         FAIL() << "should not be able to access values from fields not present in the provided models";
+      } catch (const RException &err) {
+         EXPECT_THAT(err.what(), testing::HasSubstr("invalid field name: ntuple2.z"));
+      }
+
+      ++nEntries;
+   }
+}
```

Comment:
Can you also add a test with a join operation with more than one column for the index?